### PR TITLE
Refactor to_gmfs and faster calculation method

### PR DESCRIPTION
**Pull request by schmidni**
Refactoring with the purpose of making the code extendable.
- some simple refactoring of the method to make it more readable
- added docstrings
- created CallableDict "gmfs_calculation_methods" which returns a function. This function contains the specific implementation how the gmfs are calculated, when given the matrix of mu (mean) and Z (random variables).
I could, instead of returning a function, also just return the finished result, but this way the "calculation method" and the actual "calculation" are clearly separated.

Improvement:
I also added new calculation method, which directly calculates the sampling in case no correlation is needed. Until now Cholesky was computed every time, which is pointless, if the correlation matrixes are identity matrices.

### CODE REVIEW

File: openquake/hazardlib/shakemap/gmfs.py [hunk 1]
```diff
@@ -151,41 +152,131 @@ def cholesky(spatial_cov, cross_corr):
...
+    # checks
+    N = len(shakemap)
+    M = len(imts)
+    if spatialcorr != 'no' and N > MAXSITES:
+        # hard-coded, heuristic
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE %
+                         (N, MAXSITES))
+    if N * M > cholesky_limit:
+        raise ValueError(CORRELATION_MATRIX_TOO_LARGE % (
+            '%d x %d' % (M, N), cholesky_limit))
```
Comment (schmidni) [inline openquake/hazardlib/shakemap/gmfs.py:186]: I understand the cholesky_limit check. 

But why did you add the check for the "MAXSITES" here, seeing that it is hardcoded. If it is for performance reasons when computing the spatial correlation, then the improvements from #6624 should have fixed this.

Comment (micheles) [reply openquake/hazardlib/shakemap/gmfs.py:186]: It should not be hard-coded but configurable like the cholesky_limit. Consider it at technical debt.
It is not there for performance, but to avoid small negative eigenvalues due to numerical errors. It would be nicer to have a real solution rather than the ugly workaround :-(

Comment (schmidni) [reply openquake/hazardlib/shakemap/gmfs.py:186]: But why is only checking N*M not enough?

**Review by micheles — COMMENTED** [review]
Because in the formula `numpy.linalg.cholesky(LLT)` the matrix LLT has shape (N*M, N*M)

File: openquake/hazardlib/shakemap/gmfs.py (hunk 1 above)
Comment (micheles) [reply openquake/hazardlib/shakemap/gmfs.py:186]: Because in the formula numpy.linalg.cholesky(LLT) the matrix LLT has shape `(N*M, N*M)`

Comment (schmidni) [reply openquake/hazardlib/shakemap/gmfs.py:186]: I am sorry, I think I am a bit slow.
You are checking `if N*M > cholesky_limit:` - so why do you also check `if N > MAXSITES`?

If cholesky_limit = 10000 (default) and MAXSITES = 1000 (hardcoded) then:

N = 4000 and M = 1 => ValueError
N = 999 and M = 5 => no Value Error
even though N*M is bigger in the second example.

File: openquake/hazardlib/shakemap/gmfs.py [hunk 2]
```diff
@@ -151,41 +152,131 @@ def cholesky(spatial_cov, cross_corr):
...
     spatial_cov = spatial_covariance_array(stddev, spatial_corr)
+
+    # Cholesky Decomposition
     L = cholesky(spatial_cov, cross_corr)  # shape (M * N, M * N)
+
+    # mu has unit (pctg), L has unit ln(pctg)
+    def calculate_gmfs(Z, mu):
+        mu = numpy.log(mu)
+        return numpy.exp(L @ Z + mu) / PCTG
+
```
[the remark this answers is not in the document]
Comment (schmidni) [reply openquake/hazardlib/shakemap/gmfs.py:209]: every call to `gmfs_calculation_methods` returns a function (`calculate_gmfs`). In these two cases the function looks the same (except that one has L@Z+mu and the other one sig@Z+mu). But I will add for example something along those lines:

`@gmfs_calculation_methods.add('MMI')
def gmfs_calculation_methods_basic(kind, shakemap, imts):
    # create diag matrix with std values
    std = numpy.array(shakemap['std']['MMI'])
    sig = numpy.diag(std.flatten())  # shape (M*N, M*N)

    # mu has unit (pctg), sig has unit ln(pctg)
    def calculate_gmfs(Z, mu):
        return sig @ Z + mu

    return calculate_gmfs`

File: openquake/hazardlib/shakemap/gmfs.py (hunk 1 above)
Comment (micheles) [reply openquake/hazardlib/shakemap/gmfs.py:186]: The MAXSITES check was there before, introduced years ago, I just added the additional check. Perhaps we can fully remove the check on MAXSITES and instead add a check that there are no negative eigenvalues. Thanks to the cholesky_limit we know that the decomposition will still work without running out of memory.

File: openquake/hazardlib/shakemap/gmfs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (micheles) [reply openquake/hazardlib/shakemap/gmfs.py:209]: I do not like that `gmfs_calculation_methods_basic` returns a function. It should just return the result. If you need the function you can just call `gmfs_calculation_methods_basic[kind]`. BTW, I do not like the name `gmfs_calculation_methods_basic`. It should be something like `shakemap_to_gmfs`.

[the remark this answers is not in the document]
Comment (schmidni) [reply openquake/hazardlib/shakemap/gmfs.py:208]: I don't think this helps readability and gmf_dict is huge by now since I also have to put in Z and mu. 

But you're the boss, I made the changes as requested.

File: openquake/hazardlib/shakemap/gmfs.py (hunk 1 above)
Comment (schmidni) [reply openquake/hazardlib/shakemap/gmfs.py:186]: sorry I went a bit off-topic here. I resolve this conversation and leave you to accept / ask for additional changes in the pull request.

**Review by micheles — APPROVED** [review]
